### ISSUE TITLE
Make the knobs hold instead of mash keys

### ISSUE THREAD

**Issue by okand** [5argon/UnoSDVXCon#2]
Hi!

Firstly, thank you very much for writing and sharing this code. I found a place with an SDVX arcade machine here in Stockholm, got hooked on the game and thought building something myself would be more fun than buying an SVSE5/9. Thank you for making this available.

I've been playing a bunch of k-shoot and more recently this other implementation of the game: https://github.com/Drewol/unnamed-sdvx-clone (it runs a lot better).
They did some changes to to the lasers recently which makes the keyboard controls expect holding the keys down instead of pressing them very quickly: https://github.com/Drewol/unnamed-sdvx-clone/issues/26

I tried it out on the official e-amusement cloud PC version of SDVX and had some issues there too. When turning the knobs the lasers only move a little bit by bit instead of following the tracks smoothly.

I'm pretty sure this all comes down to the Arduino repeatedly sending a character instead of holding it down. I have no idea where to even start on fixing something like this though so I thought I'd make an issue about it.

**Comment by okand** [comment]
Or maybe it's got something to do with the specific encoders I have? I got these: https://www.m.nu/knappar-avancerade-touch/rotary-encoder-20p-v-with-switch-including-washer-and-nut

No idea what manufacturer/model it is. It's got lots of small "steps" in it when you rotate it.

**Comment by 5argon** [comment]
Yes I am aware of this problem after I tried out the official e-cloud game too. In K-Shoot you can set different mode for lasers but unfortunately the official ones is expecting a hold.

Currently I could not go back to work with this, but after scanning the code for a bit you would have to understand the `void serialFromState()` function in the `.ico` file. It is designed to repeatedly press button while you are turning the knob.

HID standard can accept a total of 6 buttons per update. These buttons are filled in `buf` array which is of size 8. (You will see I start at index 2, the index 0 and 1 was reserved for something I have already forgot) And this also means if you held down 6 buttons you cannot spin any knob.

How this works is, for each time `serialFromState` is called a series of 6 numbers will be sent to the computer. If it is  [BT1, BT4, 0, 0, 0, 0] for example, it means currently button 1 and 4 is being held down and no other buttons. It's up to your computer to intepret this array on consecutive frame as button **down** or **up** or **holding**. For example 

[0,0,0,0,0,0] 
[BT1, 0, 0, 0, 0, 0] <- BT1 Down
[BT1, 0, 0, 0, 0, 0] <- **Nothing** happen from receiving this array, as the computer see no changes.
[BT4, 0, 0, 0, 0, 0] <- **BT1 Up** (because it is not present compared to the last frame) , BT4 Down
[BT1, BT4, 0, 0, 0, 0] <- BT1 Down again

If holding BT1 produce this behavior in SDVX it would be a problem as it will double press. Luckily this is not the case because as long as the button is held down the consecutive `serialFromState()` would surely add BT1 to the array. (This array will be cleared empty on each `serialFromState` call, which is when 1. button press 2. knob state change (including knob stop))

And also, if you don't send this array anymore , the computer will not assume all zeroes but still remember the last one. That is, if you unplug the controller at the last array I show the key would be held down forever, requiring pushing the computer keyboard to fix it. Please remember this rule. To say "all key up" you need to send [0,0,0,0,0,0] somehow.

If you hold down a key on your computer, it will be equivalent to having a button value in this array at all frames without any absent. Depending on application, it might results in one character typed or one character -> some delay -> repeat the character. In the gif, I turn a knob and the character appears repeatedly immediately. This means it is not a hold but rather a series of up and downs.

Now the reason why knobs produced repeating behavior is because there are (intentionally) some `buf` in between frames that lacks the VOL value.

Next, this `serialFromState` is being called by the likes of `loop`(the main loop)->`checkKnob` ->`knobLActionCW`->`setVolLState`. The important filtering stages are as follows : 

In `checkKnob` you can see `spinTolerance`. This is for preventing accident spin. Without this even when you just touch the knob it might erroneously go either way for a brief period. It is required to go over the old position for a certain range to count as a 1 spin.

Of course in between spin it will be some moment that your spin is not enough to win against this `spinTolerance`. In that case you can see it falls down to `knobL/RActionStop`. This happen even if you are not spinning too because it is in the main loop.

This is the moment that [0,0,0,0,0,0] might be sent and count as a key (knob) up. The knob state will be STOP, and then in `serialFromState` (called from both this knob spin *only if the state changed* and called from all button press, which will send the knob altogether with the button) there is no check for STOP meaning in this serialization the knob will be missing from the array. If the previous array have a knob value then the key hold will be stopped. This is what caused the repeating behavior of knob.

Initially I don't care about this hiccup as I know K-Shoot can handle hold/rapid press smoothly. But you need to modify these logic. If I would make it, after winning against the tolerance for the first time and there are frames that the spin is not enough I would

1. If the direction is continuing the same way but not enough, wait and see if there will be any movement in some time frame in the future.
2. If the direction goes the other way, check against tolerance in the opposite way. If over, immediately change spinning direction otherwise wait and see like 1.
3. If the movement stop then call the set state STOP immediately
4. If the said time frame is exceeded, also call the STOP

This will make the knob state stays at CW or CCW longer without STOP in between. Any button press while the knobs are in these state will detect the CW/CCW and send the knob key to the computer together with BT key saying that "the knob is still down". If you do not press any button, the knob would be still down as I said before no message means the last one will be in effect.

If you really manage to fix this please submit a pull request and I will merge it. Thank you!
